File: Python3/406. Queue Reconstruction by Height.py
class Solution:
    def reconstructQueue(self, people):
        """
        :type people: List[List[int]]
        :rtype: List[List[int]]
        """
        # length = len(people) # 132ms
        # ans = [[] for i in range(length)]
        # index = list(range(length))
        # personDict = {}
        # for person in people:
        #     if person[0] in personDict:
        #         personDict[person[0]].append(person[1])
        #     else:
        #         personDict[person[0]] = [person[1]]
        # for key in sorted(personDict.keys()):
        #     trashs = []
        #     for value in personDict[key]:
        #         ans[index[value]] = [key, value]
        #         trashs.append(index[value])
        #     for trash in trashs:
        #         index.remove(trash)
        # return ans
        
        # ans = [] # 104ms
        # personDict = {}
        # for person in people:
        #     if person[0] in personDict:
        #         personDict[person[0]].append(person[1])
        #     else:
        #         personDict[person[0]] = [person[1]]
        # for key in sorted(personDict.keys(), reverse=True):
        #     for value in sorted(personDict[key]):
        #         ans.insert(value, [key, value])
        # return ans
        
        ans = [] # 92ms
        people.sort(key=lambda x: (-x[0], x[1]))
        for person in people:
            ans.insert(person[1], person)
        return ans

File: Golang/406. Queue Reconstruction by Height.go
func reconstructQueue(people [][]int) [][]int { // 48ms
    res := make([][]int, 0, len(people))

    // 按照 persons 的排序方式，对 people 进行排序
    sort.Sort(persons(people))

    // 把 person 插入到 res[idx] 上
    insert := func(idx int, person []int) {
        res = append(res, person)
        // 插入到末尾
        if len(res)-1 == idx {
            return
        }
        // 插入到中间位置
        copy(res[idx+1:], res[idx:])
        res[idx] = person
    }

    for _, p := range people {
        insert(p[1], p)
    }

    return res
}

// persons 实现了 sort.Interface 接口
type persons [][]int

func (p persons) Len() int { return len(p) }

// 以 h 的降序为主
// 以 k 的升序为辅
func (p persons) Less(i, j int) bool {
    if p[i][0] == p[j][0] {
        return p[i][1] < p[j][1]
    }
    return p[i][0] > p[j][0]
}

func (p persons) Swap(i, j int) { p[i], p[j] = p[j], p[i] }